pkg/cloudprovider/providers/azure/azure_client.go

```diff
@@ -0,0 +1,811 @@
+	CreateOrUpdate(resourceGroupName string, routeTableName string, routeName string, routeParameters network.Route, cancel <-chan struct{}) (<-chan network.Route, <-chan error)
+	Delete(resourceGroupName string, routeTableName string, routeName string, cancel <-chan struct{}) (<-chan autorest.Response, <-chan error)
+}
+
+// RouteTablesClient defines needed functions for azure network.RouteTablesClient
+type RouteTablesClient interface {
+	CreateOrUpdate(resourceGroupName string, routeTableName string, parameters network.RouteTable, cancel <-chan struct{}) (<-chan network.RouteTable, <-chan error)
+	Get(resourceGroupName string, routeTableName string, expand string) (result network.RouteTable, err error)
+}
+
+// StorageAccountClient defines needed functions for azure storage.AccountsClient
+type StorageAccountClient interface {
+	Create(resourceGroupName string, accountName string, parameters storage.AccountCreateParameters, cancel <-chan struct{}) (<-chan storage.Account, <-chan error)
+	Delete(resourceGroupName string, accountName string) (result autorest.Response, err error)
+	ListKeys(resourceGroupName string, accountName string) (result storage.AccountListKeysResult, err error)
+	ListByResourceGroup(resourceGroupName string) (result storage.AccountListResult, err error)
+	GetProperties(resourceGroupName string, accountName string) (result storage.Account, err error)
+}
+
+// DisksClient defines needed functions for azure disk.DisksClient
+type DisksClient interface {
+	CreateOrUpdate(resourceGroupName string, diskName string, diskParameter disk.Model, cancel <-chan struct{}) (<-chan disk.Model, <-chan error)
+	Delete(resourceGroupName string, diskName string, cancel <-chan struct{}) (<-chan disk.OperationStatusResponse, <-chan error)
+	Get(resourceGroupName string, diskName string) (result disk.Model, err error)
+}
+
+// azVirtualMachinesClient implements VirtualMachinesClient.
+type azVirtualMachinesClient struct {
+	client      compute.VirtualMachinesClient
+	rateLimiter flowcontrol.RateLimiter
+}
+
+func newAzVirtualMachinesClient(subscriptionID string, endpoint string, servicePrincipalToken *adal.ServicePrincipalToken, rateLimiter flowcontrol.RateLimiter) *azVirtualMachinesClient {
+	virtualMachinesClient := compute.NewVirtualMachinesClient(subscriptionID)
+	virtualMachinesClient.BaseURI = endpoint
+	virtualMachinesClient.Authorizer = autorest.NewBearerAuthorizer(servicePrincipalToken)
+	virtualMachinesClient.PollingDelay = 5 * time.Second
+	configureUserAgent(&virtualMachinesClient.Client)
+
+	return &azVirtualMachinesClient{
+		rateLimiter: rateLimiter,
+		client:      virtualMachinesClient,
+	}
+}
+
+func (az *azVirtualMachinesClient) CreateOrUpdate(resourceGroupName string, VMName string, parameters compute.VirtualMachine, cancel <-chan struct{}) (<-chan compute.VirtualMachine, <-chan error) {
+	az.rateLimiter.Accept()
+	glog.V(10).Infof("azVirtualMachinesClient.CreateOrUpdate(%q, %q): start", resourceGroupName, VMName)
```

Comment:
I think it will be easier to maintain if you get name of the function dynamically using reflections. For instance with [FuncForPC](https://golang.org/pkg/runtime/#FuncForPC) and [Example](https://stackoverflow.com/questions/10742749/get-name-of-function-using-reflection-in-golang).

Reply:
@cosmincojocar Considered reflection when preparing this PR. It could really help to simplify codes, but it also introductions performance penalty. I don't think we should use it much in production.

Reply:
It wast just a suggestion. I totally agree that reflections might introduce a performance degradation.

Reply:
@cosmincojocar thanks. Let's not depend on reflection now. We may revisit this when golang has evolved high-performance reflection.